### PR TITLE
Fix missing locale in the project template's homepage migration

### LINKED ISSUE #6557: 'Create homepage' migration breaks when run after addition of locale field
### Issue Summary

The [home.0002_create_homepage](https://github.com/wagtail/wagtail/blob/master/wagtail/project_template/home/migrations/0002_create_homepage.py) migration in the project template fails if it runs after the `wagtailcore.0057_page_locale_fields_notnull` migration, since it doesn't specify a locale (and [the signal handler](https://github.com/wagtail/wagtail/blob/2cefe4ebd9f2d691e4cb963ac816ccbc62e8a697/wagtail/core/models.py#L640) doesn't fill it in, as we're running against the frozen ORM). On a new project there is no problem, because the migration engine opts to run them immediately after wagtailcore migration 0040 (due to the dependency in `home.0001_initial`), but this is not guaranteed - subsequent migrations in the 'home' app may have dependencies on later wagtailcore migrations which may cause the dependency solver to push it later in the sequence.

### Steps to Reproduce

1. Start a new project with `wagtail start myproject`
2. Edit home/models.py and change the definition of HomePage to:

       class HomePage(Page):
           other_page = models.ForeignKey(Page, null=True, blank=True, on_delete=models.SET_NULL, related_name='+')

3. Run `./manage.py makemigrations` and `./manage.py migrate`. This generates the output:

```
Operations to perform:
  Apply all migrations: admin, auth, contenttypes, home, sessions, taggit, wagtailadmin, wagtailcore, wagtaildocs, wagtailembeds, wagtailforms, wagtailimages, wagtailredirects, wagtailsearch, wagtailusers
Running migrations:
  Applying contenttypes.0001_initial... OK
  Applying auth.0001_initial... OK
  Applying admin.0001_initial... OK
  Applying admin.0002_logentry_remove_auto_add... OK
  Applying admin.0003_logentry_add_action_flag_choices... OK
  Applying contenttypes.0002_remove_content_type_name... OK
  Applying auth.0002_alter_permission_name_max_length... OK
  Applying auth.0003_alter_user_email_max_length... OK
  Applying auth.0004_alter_user_username_opts... OK
  Applying auth.0005_alter_user_last_login_null... OK
  Applying auth.0006_require_contenttypes_0002... OK
  Applying auth.0007_alter_validators_add_error_messages... OK
  Applying auth.0008_alter_user_username_max_length... OK
  Applying auth.0009_alter_user_last_name_max_length... OK
  Applying auth.0010_alter_group_name_max_length... OK
  Applying auth.0011_update_proxy_permissions... OK
  Applying wagtailcore.0001_squashed_0016_change_page_url_path_to_text_field... OK
  Applying wagtailcore.0017_change_edit_page_permission_description... OK
  Applying wagtailcore.0018_pagerevision_submitted_for_moderation_index... OK
  Applying wagtailcore.0019_verbose_names_cleanup... OK
  Applying wagtailcore.0020_add_index_on_page_first_published_at... OK
  Applying wagtailcore.0021_capitalizeverbose... OK
  Applying wagtailcore.0022_add_site_name... OK
  Applying wagtailcore.0023_alter_page_revision_on_delete_behaviour... OK
  Applying wagtailcore.0024_collection... OK
  Applying wagtailcore.0025_collection_initial_data... OK
  Applying wagtailcore.0026_group_collection_permission... OK
  Applying wagtailcore.0027_fix_collection_path_collation... OK
  Applying wagtailcore.0024_alter_page_content_type_on_delete_behaviour... OK
  Applying wagtailcore.0028_merge... OK
  Applying wagtailcore.0029_unicode_slugfield_dj19... OK
  Applying wagtailcore.0030_index_on_pagerevision_created_at... OK
  Applying wagtailcore.0031_add_page_view_restriction_types... OK
  Applying wagtailcore.0032_add_bulk_delete_page_permission... OK
  Applying wagtailcore.0033_remove_golive_expiry_help_text... OK
  Applying wagtailcore.0034_page_live_revision... OK
  Applying wagtailcore.0035_page_last_published_at... OK
  Applying wagtailcore.0036_populate_page_last_published_at... OK
  Applying wagtailcore.0037_set_page_owner_editable... OK
  Applying wagtailcore.0038_make_first_published_at_editable... OK
  Applying wagtailcore.0039_collectionviewrestriction... OK
  Applying wagtailcore.0040_page_draft_title... OK
  Applying wagtailcore.0041_group_collection_permissions_verbose_name_plural... OK
  Applying wagtailcore.0042_index_on_pagerevision_approved_go_live_at... OK
  Applying wagtailcore.0043_lock_fields... OK
  Applying wagtailcore.0044_add_unlock_grouppagepermission... OK
  Applying wagtailcore.0045_assign_unlock_grouppagepermission... OK
  Applying wagtailcore.0046_site_name_remove_null... OK
  Applying wagtailcore.0047_add_workflow_models... OK
  Applying wagtailcore.0048_add_default_workflows... OK
  Applying wagtailcore.0049_taskstate_finished_by... OK
  Applying wagtailcore.0050_workflow_rejected_to_needs_changes... OK
  Applying wagtailcore.0051_taskstate_comment... OK
  Applying wagtailcore.0052_pagelogentry... OK
  Applying wagtailcore.0053_locale_model... OK
  Applying wagtailcore.0054_initial_locale... OK
  Applying wagtailcore.0055_page_locale_fields... OK
  Applying wagtailcore.0056_page_locale_fields_populate... OK
  Applying wagtailcore.0057_page_locale_fields_notnull... OK
  Applying wagtailcore.0058_page_alias_of... OK
  Applying wagtailcore.0059_apply_collection_ordering... OK
  Applying home.0001_initial... OK
  Applying home.0002_create_homepage...Traceback (most recent call last):
  File "/home/vagrant/.virtualenvs/bakerydemo/lib/python3.6/site-packages/django/db/backends/utils.py", line 86, in _execute
    return self.cursor.execute(sql, params)
  File "/home/vagrant/.virtualenvs/bakerydemo/lib/python3.6/site-packages/django/db/backends/sqlite3/base.py", line 396, in execute
    return Database.Cursor.execute(self, query, params)
sqlite3.IntegrityError: NOT NULL constraint failed: wagtailcore_page.locale_id

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/vagrant/.virtualenvs/bakerydemo/lib/python3.6/site-packages/django/db/migrations/executor.py", line 245, in apply_migration
    state = migration.apply(state, schema_editor)
  File "/home/vagrant/.virtualenvs/bakerydemo/lib/python3.6/site-packages/django/db/migrations/migration.py", line 124, in apply
    operation.database_forwards(self.app_label, schema_editor, old_state, project_state)
  File "/home/vagrant/.virtualenvs/bakerydemo/lib/python3.6/site-packages/django/db/migrations/operations/special.py", line 190, in database_forwards
    self.code(from_state.apps, schema_editor)
  File "/vagrant/libs/debug/migrationorder/home/migrations/0002_create_homepage.py", line 29, in create_homepage
    url_path='/home/',
  File "/home/vagrant/.virtualenvs/bakerydemo/lib/python3.6/site-packages/django/db/models/manager.py", line 82, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/home/vagrant/.virtualenvs/bakerydemo/lib/python3.6/site-packages/django/db/models/query.py", line 433, in create
    obj.save(force_insert=True, using=self.db)
  File "/home/vagrant/.virtualenvs/bakerydemo/lib/python3.6/site-packages/django/db/models/base.py", line 749, in save
    force_update=force_update, update_fields=update_fields)
  File "/home/vagrant/.virtualenvs/bakerydemo/lib/python3.6/site-packages/django/db/models/base.py", line 784, in save_base
    parent_inserted = self._save_parents(cls, using, update_fields)
  File "/home/vagrant/.virtualenvs/bakerydemo/lib/python3.6/site-packages/django/db/models/base.py", line 815, in _save_parents
    force_insert=parent_inserted,
  File "/home/vagrant/.virtualenvs/bakerydemo/lib/python3.6/site-packages/django/db/models/base.py", line 890, in _save_table
    results = self._do_insert(cls._base_manager, using, fields, returning_fields, raw)
  File "/home/vagrant/.virtualenvs/bakerydemo/lib/python3.6/site-packages/django/db/models/base.py", line 929, in _do_insert
    using=using, raw=raw,
  File "/home/vagrant/.virtualenvs/bakerydemo/lib/python3.6/site-packages/django/db/models/manager.py", line 82, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/home/vagrant/.virtualenvs/bakerydemo/lib/python3.6/site-packages/django/db/models/query.py", line 1204, in _insert
    return query.get_compiler(using=using).execute_sql(returning_fields)
  File "/home/vagrant/.virtualenvs/bakerydemo/lib/python3.6/site-packages/django/db/models/sql/compiler.py", line 1394, in execute_sql
    cursor.execute(sql, params)
  File "/home/vagrant/.virtualenvs/bakerydemo/lib/python3.6/site-packages/django/db/backends/utils.py", line 100, in execute
    return super().execute(sql, params)
  File "/home/vagrant/.virtualenvs/bakerydemo/lib/python3.6/site-packages/django/db/backends/utils.py", line 68, in execute
    return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)
  File "/home/vagrant/.virtualenvs/bakerydemo/lib/python3.6/site-packages/django/db/backends/utils.py", line 77, in _execute_with_wrappers
    return executor(sql, params, many, context)
  File "/home/vagrant/.virtualenvs/bakerydemo/lib/python3.6/site-packages/django/db/backends/utils.py", line 86, in _execute
    return self.cursor.execute(sql, params)
  File "/home/vagrant/.virtualenvs/bakerydemo/lib/python3.6/site-packages/django/db/utils.py", line 90, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/home/vagrant/.virtualenvs/bakerydemo/lib/python3.6/site-packages/django/db/backends/utils.py", line 86, in _execute
    return self.cursor.execute(sql, params)
  File "/home/vagrant/.virtualenvs/bakerydemo/lib/python3.6/site-packages/django/db/backends/sqlite3/base.py", line 396, in execute
    return Database.Cursor.execute(self, query, params)
django.db.utils.IntegrityError: NOT NULL constraint failed: wagtailcore_page.locale_id

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "./manage.py", line 10, in <module>
    execute_from_command_line(sys.argv)
  File "/home/vagrant/.virtualenvs/bakerydemo/lib/python3.6/site-packages/django/core/management/__init__.py", line 401, in execute_from_command_line
    utility.execute()
  File "/home/vagrant/.virtualenvs/bakerydemo/lib/python3.6/site-packages/django/core/management/__init__.py", line 395, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/home/vagrant/.virtualenvs/bakerydemo/lib/python3.6/site-packages/django/core/management/base.py", line 328, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/home/vagrant/.virtualenvs/bakerydemo/lib/python3.6/site-packages/django/core/management/base.py", line 369, in execute
    output = self.handle(*args, **options)
  File "/home/vagrant/.virtualenvs/bakerydemo/lib/python3.6/site-packages/django/core/management/base.py", line 83, in wrapped
    res = handle_func(*args, **kwargs)
  File "/home/vagrant/.virtualenvs/bakerydemo/lib/python3.6/site-packages/django/core/management/commands/migrate.py", line 233, in handle
    fake_initial=fake_initial,
  File "/home/vagrant/.virtualenvs/bakerydemo/lib/python3.6/site-packages/django/db/migrations/executor.py", line 117, in migrate
    state = self._migrate_all_forwards(state, plan, full_plan, fake=fake, fake_initial=fake_initial)
  File "/home/vagrant/.virtualenvs/bakerydemo/lib/python3.6/site-packages/django/db/migrations/executor.py", line 147, in _migrate_all_forwards
    state = self.apply_migration(state, migration, fake=fake, fake_initial=fake_initial)
  File "/home/vagrant/.virtualenvs/bakerydemo/lib/python3.6/site-packages/django/db/migrations/executor.py", line 247, in apply_migration
    migration_recorded = True
  File "/home/vagrant/.virtualenvs/bakerydemo/lib/python3.6/site-packages/django/db/backends/sqlite3/schema.py", line 35, in __exit__
    self.connection.check_constraints()
  File "/home/vagrant/.virtualenvs/bakerydemo/lib/python3.6/site-packages/django/db/backends/sqlite3/base.py", line 307, in check_constraints
    violations = self.cursor().execute('PRAGMA foreign_key_check').fetchall()
  File "/home/vagrant/.virtualenvs/bakerydemo/lib/python3.6/site-packages/django/db/backends/utils.py", line 100, in execute
    return super().execute(sql, params)
  File "/home/vagrant/.virtualenvs/bakerydemo/lib/python3.6/site-packages/django/db/backends/utils.py", line 68, in execute
    return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)
  File "/home/vagrant/.virtualenvs/bakerydemo/lib/python3.6/site-packages/django/db/backends/utils.py", line 77, in _execute_with_wrappers
    return executor(sql, params, many, context)
  File "/home/vagrant/.virtualenvs/bakerydemo/lib/python3.6/site-packages/django/db/backends/utils.py", line 80, in _execute
    self.db.validate_no_broken_transaction()
  File "/home/vagrant/.virtualenvs/bakerydemo/lib/python3.6/site-packages/django/db/backends/base/base.py", line 449, in validate_no_broken_transaction
    "An error occurred in the current transaction. You can't "
django.db.transaction.TransactionManagementError: An error occurred in the current transaction. You can't execute queries until the end of the 'atomic' block.
```

Any other relevant information. For example, why do you consider this a bug and what did you expect to happen instead?

* I have confirmed that this issue can be reproduced as described on a fresh Wagtail project: yes

### Technical details

* Python version: 3.6.9
* Django version: 3.0.10
* Wagtail version: master (2.12a0)

### PR DESCRIPTION
**Pull request by zerolab**
Fixes #6557

Added "upgrade considerations", which may be a bit on the controversial side so input is more than welcome

### CODE REVIEW

**Comment by gasman** [comment]
Thanks @zerolab!
I think our advice to upgrading users should be to add `run_before = ('wagtailcore', '0053_locale_model')` to the existing migration, so that it runs in the same position in the sequence as before.

Updating the project template to work against the post-locale version does intuitively feel like the right thing to do for new projects, for the sake of neatness - but the more I think about it, the more I feel like we'd be better off playing it safe and keeping this `run_before` fix in place. From experience, the best way to manage complexity in migrations is to make as few retrospective changes as possible, to minimise the number of variations of environments that exist in the wild.

**Comment by zerolab** [comment]
>  I think our advice to upgrading users should be to add `run_before = ('wagtailcore', '0053_locale_model')` to the existing migration, so that it runs in the same position in the sequence as before.

Will do so for the homepage migration. My concern with that is could break in other ways in other user custom migrations  (i.e. create this "TeamIndex" page programatically) after we add further migrations, say in 2.12. Happy to tone it down and keep it simple, though

**Comment by gasman** [comment]
> My concern with that is could break in other ways in other user custom migrations (i.e. create this "TeamIndex" page programatically) after we add further migrations, say in 2.12

I think that once we have a `run_before` line in place on any migration that creates pages, we should be insured against any future changes to the Page model.

**Comment by zerolab** [comment]
Updated.

File: wagtail/project_template/home/migrations/0002_create_homepage.py [hunk 1]
```diff
@@ -49,6 +49,8 @@ def remove_homepage(apps, schema_editor):
 
 class Migration(migrations.Migration):
 
+    run_before = ('wagtailcore', '0053_locale_model')
```
Comment (dannyhajj) [inline wagtail/project_template/home/migrations/0002_create_homepage.py:52]: `run_before` should be an [array of tuples](https://docs.djangoproject.com/en/3.1/howto/writing-migrations/#controlling-the-order-of-migrations). As it is now, Django will raise a `NodeNotFoundError` exception with a `nonexistent child node 'wagtailcore'` message.

```
run_before = [
    ('wagtailcore', '0053_locale_model'),
]
```

Comment (zerolab) [reply wagtail/project_template/home/migrations/0002_create_homepage.py]: Thank you for taking the time to review this. You are right. Should be a lesson not to make tweaks in a rush (busy schedule be damned) 🤦‍♂️

**Review by dannyhajj — APPROVED** [review]

File: docs/releases/2.11.3.rst [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+~~~~~~~~~
+
+ *
+
+
+Upgrade considerations
+======================
+
+If you have used the Wagtail project template, or have migrations that create page instances programmatically, you will
+need to add ``run_before = [('wagtailcore', '0053_locale_model')]`` to the corresponding ``Migration`` class.
```
Comment (stevejalim) [inline docs/releases/2.11.3.rst:23]: @zerolab Is it worth expanding on what will happen if you do not? Providing an example error message may help with the discovery of this updgrade tip

File: docs/releases/2.11.3.rst [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+What's new
+==========
+
+Bug fixes
+~~~~~~~~~
+
+ *
+
+
+Upgrade considerations
```
Comment (kaedroho) [inline docs/releases/2.11.3.rst:19]: I don't usually look in point releases for upgrade considerations, would it be worth adding a link to this from the 2.11 upgrade considerations for those upgrading later?

Comment (gasman) [reply docs/releases/2.11.3.rst:19]: Yep, I agree.

**Comment by gasman** [comment]
Merged in 0da2ba69bcf971f481e00dc8e6e374888e3517bb (master) and 3e24b5b398d96ab259b2b03d59cb6e31a52ca7d1 (stable/2.11.x), with further details added to the release notes in 9a953fd93cccb8288d4399d94240976d3a14e2fa / 27758937a6a2173c713008f27c840b0c382c3fe7 to cover @kaedroho and @stevejalim's suggestions.

(further feedback on my doc changes is welcome - just wanted to get it merged before things got messy with merge conflicts due to the release notes page being added here...)

**Comment by camilolucena88** [comment]
I have an issue that after add the run_before, I will get this:

``raise CircularDependencyError(", ".join("%s.%s" % n for n in cycle))
django.db.migrations.exceptions.CircularDependencyError: wagtailcore.0053_locale_model, wagtailcore.0054_initial_locale, wagtailcore.0055_page_locale_fields, wagtailcore.0056_page_locale_fields_populate, wagtailcore.0057_page_locale_fields_notnull, wagtailcore.0058_page_alias_of, wagtailcore.0059_apply_collection_ordering, home.0001_initial, home.0002_create_homepage``

Please, if someone can give me a hand on this one

**Comment by gasman** [comment]
@camilolucena88 Did you make any changes to the `home/0001_initial` migration? The `dependencies` line on that migration should remain as

    dependencies = [
        ('wagtailcore', '0040_page_draft_title'),
    ]

so that it's valid for it to run before the new migrations from 2.11.

**Comment by camilolucena88** [comment]
Thanks for your fast reply. It was solved with that. I had a few changes on that file and I didn't notice it.

**Pull request closed without merging**